Make connections retry in more spots

<details>
<summary>eg. this one -- should not just fail, but rather retry a number of times. So we might want to provision it somehow more globally</summary> 

```shell
=================================== FAILURES ===================================
_ test_parse_api_url[https://gui.dandiarchive.org/#/dandiset/000001-parsed_url0] _
/opt/hostedtoolcache/Python/3.10.16/x64/lib/python3.10/site-packages/urllib3/connectionpool.py:787: in urlopen
    response = self._make_request(
/opt/hostedtoolcache/Python/3.10.16/x64/lib/python3.10/site-packages/urllib3/connectionpool.py:488: in _make_request
    raise new_e
/opt/hostedtoolcache/Python/3.10.16/x64/lib/python3.10/site-packages/urllib3/connectionpool.py:464: in _make_request
    self._validate_conn(conn)
/opt/hostedtoolcache/Python/3.10.16/x64/lib/python3.10/site-packages/urllib3/connectionpool.py:1093: in _validate_conn
    conn.connect()
/opt/hostedtoolcache/Python/3.10.16/x64/lib/python3.10/site-packages/urllib3/connection.py:741: in connect
    sock_and_verified = _ssl_wrap_socket_and_match_hostname(
/opt/hostedtoolcache/Python/3.10.16/x64/lib/python3.10/site-packages/urllib3/connection.py:920: in _ssl_wrap_socket_and_match_hostname
    ssl_sock = ssl_wrap_socket(
/opt/hostedtoolcache/Python/3.10.16/x64/lib/python3.10/site-packages/urllib3/util/ssl_.py:460: in ssl_wrap_socket
    ssl_sock = _ssl_wrap_socket_impl(sock, context, tls_in_tls, server_hostname)
/opt/hostedtoolcache/Python/3.10.16/x64/lib/python3.10/site-packages/urllib3/util/ssl_.py:504: in _ssl_wrap_socket_impl
    return ssl_context.wrap_socket(sock, server_hostname=server_hostname)
/opt/hostedtoolcache/Python/3.10.16/x64/lib/python3.10/ssl.py:513: in wrap_socket
    return self.sslsocket_class._create(
/opt/hostedtoolcache/Python/3.10.16/x64/lib/python3.10/ssl.py:1104: in _create
    self.do_handshake()
/opt/hostedtoolcache/Python/3.10.16/x64/lib/python3.10/ssl.py:1375: in do_handshake
    self._sslobj.do_handshake()
E   ConnectionResetError: [Errno 104] Connection reset by peer

During handling of the above exception, another exception occurred:
/opt/hostedtoolcache/Python/3.10.16/x64/lib/python3.10/site-packages/requests/adapters.py:667: in send
    resp = conn.urlopen(
/opt/hostedtoolcache/Python/3.10.16/x64/lib/python3.10/site-packages/urllib3/connectionpool.py:841: in urlopen
    retries = retries.increment(
/opt/hostedtoolcache/Python/3.10.16/x64/lib/python3.10/site-packages/urllib3/util/retry.py:474: in increment
    raise reraise(type(error), error, _stacktrace)
/opt/hostedtoolcache/Python/3.10.16/x64/lib/python3.10/site-packages/urllib3/util/util.py:38: in reraise
    raise value.with_traceback(tb)
/opt/hostedtoolcache/Python/3.10.16/x64/lib/python3.10/site-packages/urllib3/connectionpool.py:787: in urlopen
    response = self._make_request(
/opt/hostedtoolcache/Python/3.10.16/x64/lib/python3.10/site-packages/urllib3/connectionpool.py:488: in _make_request
    raise new_e
/opt/hostedtoolcache/Python/3.10.16/x64/lib/python3.10/site-packages/urllib3/connectionpool.py:464: in _make_request
    self._validate_conn(conn)
/opt/hostedtoolcache/Python/3.10.16/x64/lib/python3.10/site-packages/urllib3/connectionpool.py:1093: in _validate_conn
    conn.connect()
/opt/hostedtoolcache/Python/3.10.16/x64/lib/python3.10/site-packages/urllib3/connection.py:741: in connect
    sock_and_verified = _ssl_wrap_socket_and_match_hostname(
/opt/hostedtoolcache/Python/3.10.16/x64/lib/python3.10/site-packages/urllib3/connection.py:920: in _ssl_wrap_socket_and_match_hostname
    ssl_sock = ssl_wrap_socket(
/opt/hostedtoolcache/Python/3.10.16/x64/lib/python3.10/site-packages/urllib3/util/ssl_.py:460: in ssl_wrap_socket
    ssl_sock = _ssl_wrap_socket_impl(sock, context, tls_in_tls, server_hostname)
/opt/hostedtoolcache/Python/3.10.16/x64/lib/python3.10/site-packages/urllib3/util/ssl_.py:504: in _ssl_wrap_socket_impl
    return ssl_context.wrap_socket(sock, server_hostname=server_hostname)
/opt/hostedtoolcache/Python/3.10.16/x64/lib/python3.10/ssl.py:513: in wrap_socket
    return self.sslsocket_class._create(
/opt/hostedtoolcache/Python/3.10.16/x64/lib/python3.10/ssl.py:1104: in _create
    self.do_handshake()
/opt/hostedtoolcache/Python/3.10.16/x64/lib/python3.10/ssl.py:1375: in do_handshake
    self._sslobj.do_handshake()
E   urllib3.exceptions.ProtocolError: ('Connection aborted.', ConnectionResetError(104, 'Connection reset by peer'))

During handling of the above exception, another exception occurred:
dandi/tests/test_dandiarchive.py:358: in test_parse_api_url
    assert parse_dandi_url(url) == parsed_url
dandi/dandiarchive.py:755: in parse
    new_url = cls.follow_redirect(url)
dandi/dandiarchive.py:894: in follow_redirect
    r = requests.head(url, allow_redirects=True)
/opt/hostedtoolcache/Python/3.10.16/x64/lib/python3.10/site-packages/requests/api.py:100: in head
    return request("head", url, **kwargs)
/opt/hostedtoolcache/Python/3.10.16/x64/lib/python3.10/site-packages/requests/api.py:59: in request
    return session.request(method=method, url=url, **kwargs)
/opt/hostedtoolcache/Python/3.10.16/x64/lib/python3.10/site-packages/requests/sessions.py:589: in request
    resp = self.send(prep, **send_kwargs)
/opt/hostedtoolcache/Python/3.10.16/x64/lib/python3.10/site-packages/requests/sessions.py:724: in send
    history = [resp for resp in gen]
/opt/hostedtoolcache/Python/3.10.16/x64/lib/python3.10/site-packages/requests/sessions.py:724: in <listcomp>
    history = [resp for resp in gen]
/opt/hostedtoolcache/Python/3.10.16/x64/lib/python3.10/site-packages/requests/sessions.py:265: in resolve_redirects
    resp = self.send(
/opt/hostedtoolcache/Python/3.10.16/x64/lib/python3.10/site-packages/requests/sessions.py:703: in send
    r = adapter.send(request, **kwargs)
/opt/hostedtoolcache/Python/3.10.16/x64/lib/python3.10/site-packages/requests/adapters.py:682: in send
    raise ConnectionError(err, request=request)
E   requests.exceptions.ConnectionError: ('Connection aborted.', ConnectionResetError(104, 'Connection reset by peer'))
```
</details>

@jwodder do you see the best way to do that overall at requests level to retry all connection resets?

--- a/dandi/dandiarchive.py
+++ b/dandi/dandiarchive.py
@@ -889,31 +889,29 @@ class _dandi_url_parser:
         :raises FailedToConnectError: if a response other than 200, 400, 404,
             or one of the statuses in `~dandi.consts.RETRY_STATUSES` is returned
         """
-        i = 0
         retries = 4
-        while True:
+        for retry in range(retries + 1):
             try:
                 r = requests.head(url, allow_redirects=True)
             except requests.ConnectionError as e:
                 # we frequently get those on Windows for some reason
-                if i < retries:
+                if retry < retries:
                     lgr.warning(
                         "HEAD request to %s failed with %s; retrying...",
                         url,
                         e,
                     )
-                    sleep(0.1 * 10**i)
-                    i += 1
+                    sleep(0.1 * 10**retry)
                     continue
                 else:
                     lgr.warning("Exhausted %d retries for %s", retries, url)
                     raise  # re-raise the exception
-            if r.status_code in RETRY_STATUSES and i < retries:
+            if r.status_code in RETRY_STATUSES and retry < retries:
                 retry_after = get_retry_after(r)
                 if retry_after is not None:
                     delay = retry_after
                 else:
-                    delay = 0.1 * 10**i
+                    delay = 0.1 * 10**retry
                 if delay:
                     lgr.warning(
                         "HEAD request to %s returned %d; "
@@ -929,7 +927,6 @@ class _dandi_url_parser:
                         url,
                         r.status_code,
                     )
-                i += 1
                 continue
             elif r.status_code == 404:
                 raise NotFoundError(url)

--- a/dandi/dandiarchive.py
+++ b/dandi/dandiarchive.py
@@ -940,6 +940,7 @@ class _dandi_url_parser:
                 return r.url
             assert isinstance(url, str)
             return url
+        raise RuntimeError("Must not get here: either return or raise in the loop")
 
 
 # convenience binding

--- a/dandi/dandiarchive.py
+++ b/dandi/dandiarchive.py
@@ -890,9 +890,25 @@ class _dandi_url_parser:
             or one of the statuses in `~dandi.consts.RETRY_STATUSES` is returned
         """
         i = 0
+        retries = 4
         while True:
-            r = requests.head(url, allow_redirects=True)
-            if r.status_code in RETRY_STATUSES and i < 4:
+            try:
+                r = requests.head(url, allow_redirects=True)
+            except requests.ConnectionError as e:
+                # we frequently get those on Windows for some reason
+                if i < retries:
+                    lgr.warning(
+                        "HEAD request to %s failed with %s; retrying...",
+                        url,
+                        e,
+                    )
+                    sleep(0.1 * 10**i)
+                    i += 1
+                    continue
+                else:
+                    lgr.warning("Exhausted %d retries for %s", retries, url)
+                    raise  # re-raise the exception
+            if r.status_code in RETRY_STATUSES and i < retries:
                 retry_after = get_retry_after(r)
                 if retry_after is not None:
                     delay = retry_after

--- a/dandi/tests/test_dandiarchive.py
+++ b/dandi/tests/test_dandiarchive.py
@@ -28,16 +28,6 @@ from .fixtures import DandiAPI, SampleDandiset
 @pytest.mark.parametrize(
     "url,parsed_url",
     [
-        # DANDI web UI driven by DANDI API.
-        pytest.param(
-            "https://dandiarchive.org/dandiset/000001",
-            DandisetURL(
-                instance=known_instances["dandi"],
-                dandiset_id="000001",
-                version_id=None,
-            ),
-            marks=mark.skipif_no_network,
-        ),
         # Test the one with /#/ in URL (redirected)
         pytest.param(
             "https://dandiarchive.org/#/dandiset/000001",
@@ -48,8 +38,9 @@ from .fixtures import DandiAPI, SampleDandiset
             ),
             marks=mark.skipif_no_network,
         ),
+        # DANDI web UI driven by DANDI API.
         pytest.param(
-            "https://dandiarchive.org/dandiset/000001/",
+            "https://dandiarchive.org/dandiset/000001",
             DandisetURL(
                 instance=known_instances["dandi"],
                 dandiset_id="000001",
@@ -68,42 +59,6 @@ from .fixtures import DandiAPI, SampleDandiset
         ),
         pytest.param(
             "https://dandiarchive.org/dandiset/000001/0.201104.2302/",
-            DandisetURL(
-                instance=known_instances["dandi"],
-                dandiset_id="000001",
-                version_id="0.201104.2302",
-            ),
-            marks=mark.skipif_no_network,
-        ),
-        pytest.param(
-            "https://dandiarchive.org/dandiset/000001/0.201104.2302/files",
-            DandisetURL(
-                instance=known_instances["dandi"],
-                dandiset_id="000001",
-                version_id="0.201104.2302",
-            ),
-            marks=mark.skipif_no_network,
-        ),
-        pytest.param(
-            "https://dandiarchive.org/dandiset/000001/draft",
-            DandisetURL(
-                instance=known_instances["dandi"],
-                dandiset_id="000001",
-                version_id="draft",
-            ),
-            marks=mark.skipif_no_network,
-        ),
-        pytest.param(
-            "https://dandiarchive.org/dandiset/000001",
-            DandisetURL(
-                instance=known_instances["dandi"],
-                dandiset_id="000001",
-                version_id=None,
-            ),
-            marks=mark.skipif_no_network,
-        ),
-        pytest.param(
-            "https://dandiarchive.org/dandiset/000001/0.201104.2302",
             DandisetURL(
                 instance=known_instances["dandi"],
                 dandiset_id="000001",
@@ -170,7 +125,7 @@ from .fixtures import DandiAPI, SampleDandiset
                 version_id=None,
             ),
         ),
-        (
+        (  # no trailing /
             "http://localhost:8000/api/dandisets/000002",
             DandisetURL(
                 instance=known_instances["dandi-api-local-docker-tests"],

--- a/dandi/tests/test_dandiarchive.py
+++ b/dandi/tests/test_dandiarchive.py
@@ -1,6 +1,8 @@
 import re
+from unittest.mock import patch
 
 import pytest
+import requests
 import responses
 
 from dandi.consts import DandiInstance, known_instances
@@ -448,6 +450,46 @@ def test_parse_dandi_url_not_found() -> None:
 def test_follow_redirect() -> None:
     url = follow_redirect("https://bit.ly/dandi12")
     assert re.match(r"https://(.*\.)?dandiarchive.org", url)
+
+
+def test_follow_redirect_retry_on_connection_error() -> None:
+    """Test that follow_redirect retries on requests.ConnectionError and eventually succeeds."""
+    test_url = "https://example.com/test"
+    expected_url = "https://example.com/redirected"
+
+    # Mock requests.head to raise ConnectionError first 2 times, then succeed
+    call_count = 0
+
+    def mock_head(url, allow_redirects=True):
+        nonlocal call_count
+        call_count += 1
+        if call_count <= 2:
+            raise requests.ConnectionError("Connection failed")
+        # On third call, return successful response
+        response = requests.Response()
+        response.status_code = 200
+        response.url = expected_url
+        return response
+
+    with patch("dandi.dandiarchive.requests.head", side_effect=mock_head):
+        with patch("dandi.dandiarchive.sleep"):  # Mock sleep to speed up test
+            result = follow_redirect(test_url)
+            assert result == expected_url
+            assert call_count == 3  # Should have retried 2 times before succeeding
+
+
+def test_follow_redirect_exhausted_retries_on_connection_error() -> None:
+    """Test that follow_redirect raises ConnectionError after exhausting retries."""
+    test_url = "https://example.com/test"
+
+    # Mock requests.head to always raise ConnectionError
+    def mock_head(url, allow_redirects=True):
+        raise requests.ConnectionError("Connection failed")
+
+    with patch("dandi.dandiarchive.requests.head", side_effect=mock_head):
+        with patch("dandi.dandiarchive.sleep"):  # Mock sleep to speed up test
+            with pytest.raises(requests.ConnectionError, match="Connection failed"):
+                follow_redirect(test_url)
 
 
 @responses.activate

--- a/dandi/tests/test_dandiarchive.py
+++ b/dandi/tests/test_dandiarchive.py
@@ -28,7 +28,7 @@ from .fixtures import DandiAPI, SampleDandiset
 @pytest.mark.parametrize(
     "url,parsed_url",
     [
-        # New DANDI web UI driven by DANDI API.
+        # DANDI web UI driven by DANDI API.
         pytest.param(
             "https://dandiarchive.org/dandiset/000001",
             DandisetURL(

--- a/dandi/tests/test_dandiarchive.py
+++ b/dandi/tests/test_dandiarchive.py
@@ -28,7 +28,17 @@ from .fixtures import DandiAPI, SampleDandiset
     [
         # New DANDI web UI driven by DANDI API.
         pytest.param(
-            "https://gui.dandiarchive.org/#/dandiset/000001",
+            "https://dandiarchive.org/dandiset/000001",
+            DandisetURL(
+                instance=known_instances["dandi"],
+                dandiset_id="000001",
+                version_id=None,
+            ),
+            marks=mark.skipif_no_network,
+        ),
+        # Test the one with /#/ in URL (redirected)
+        pytest.param(
+            "https://dandiarchive.org/#/dandiset/000001",
             DandisetURL(
                 instance=known_instances["dandi"],
                 dandiset_id="000001",
@@ -37,7 +47,7 @@ from .fixtures import DandiAPI, SampleDandiset
             marks=mark.skipif_no_network,
         ),
         pytest.param(
-            "https://gui.dandiarchive.org/#/dandiset/000001/",
+            "https://dandiarchive.org/dandiset/000001/",
             DandisetURL(
                 instance=known_instances["dandi"],
                 dandiset_id="000001",
@@ -46,7 +56,7 @@ from .fixtures import DandiAPI, SampleDandiset
             marks=mark.skipif_no_network,
         ),
         pytest.param(
-            "https://gui.dandiarchive.org/#/dandiset/000001/0.201104.2302",
+            "https://dandiarchive.org/dandiset/000001/0.201104.2302",
             DandisetURL(
                 instance=known_instances["dandi"],
                 dandiset_id="000001",
@@ -55,7 +65,7 @@ from .fixtures import DandiAPI, SampleDandiset
             marks=mark.skipif_no_network,
         ),
         pytest.param(
-            "https://gui.dandiarchive.org/#/dandiset/000001/0.201104.2302/",
+            "https://dandiarchive.org/dandiset/000001/0.201104.2302/",
             DandisetURL(
                 instance=known_instances["dandi"],
                 dandiset_id="000001",
@@ -64,7 +74,7 @@ from .fixtures import DandiAPI, SampleDandiset
             marks=mark.skipif_no_network,
         ),
         pytest.param(
-            "https://gui.dandiarchive.org/#/dandiset/000001/0.201104.2302/files",
+            "https://dandiarchive.org/dandiset/000001/0.201104.2302/files",
             DandisetURL(
                 instance=known_instances["dandi"],
                 dandiset_id="000001",
@@ -73,7 +83,7 @@ from .fixtures import DandiAPI, SampleDandiset
             marks=mark.skipif_no_network,
         ),
         pytest.param(
-            "https://gui.dandiarchive.org/#/dandiset/000001/draft",
+            "https://dandiarchive.org/dandiset/000001/draft",
             DandisetURL(
                 instance=known_instances["dandi"],
                 dandiset_id="000001",
@@ -82,7 +92,7 @@ from .fixtures import DandiAPI, SampleDandiset
             marks=mark.skipif_no_network,
         ),
         pytest.param(
-            "https://gui.dandiarchive.org/dandiset/000001",
+            "https://dandiarchive.org/dandiset/000001",
             DandisetURL(
                 instance=known_instances["dandi"],
                 dandiset_id="000001",
@@ -91,7 +101,7 @@ from .fixtures import DandiAPI, SampleDandiset
             marks=mark.skipif_no_network,
         ),
         pytest.param(
-            "https://gui.dandiarchive.org/dandiset/000001/0.201104.2302",
+            "https://dandiarchive.org/dandiset/000001/0.201104.2302",
             DandisetURL(
                 instance=known_instances["dandi"],
                 dandiset_id="000001",
@@ -100,7 +110,7 @@ from .fixtures import DandiAPI, SampleDandiset
             marks=mark.skipif_no_network,
         ),
         pytest.param(
-            "https://gui.dandiarchive.org/dandiset/000001/0.201104.2302/files",
+            "https://dandiarchive.org/dandiset/000001/0.201104.2302/files",
             DandisetURL(
                 instance=known_instances["dandi"],
                 dandiset_id="000001",
@@ -109,7 +119,7 @@ from .fixtures import DandiAPI, SampleDandiset
             marks=mark.skipif_no_network,
         ),
         pytest.param(
-            "https://gui.dandiarchive.org/dandiset/000001/draft",
+            "https://dandiarchive.org/dandiset/000001/draft",
             DandisetURL(
                 instance=known_instances["dandi"],
                 dandiset_id="000001",
@@ -191,7 +201,7 @@ from .fixtures import DandiAPI, SampleDandiset
             ),
         ),
         pytest.param(
-            "https://gui.dandiarchive.org/#/dandiset/000001/files"
+            "https://dandiarchive.org/#/dandiset/000001/files"
             "?location=%2Fsub-anm369962",
             AssetFolderURL(
                 instance=known_instances["dandi"],
@@ -202,7 +212,7 @@ from .fixtures import DandiAPI, SampleDandiset
             marks=mark.skipif_no_network,
         ),
         pytest.param(
-            "https://gui.dandiarchive.org/#/dandiset/000006/0.200714.1807/files"
+            "https://dandiarchive.org/#/dandiset/000006/0.200714.1807/files"
             "?location=%2Fsub-anm369962",
             AssetFolderURL(
                 instance=known_instances["dandi"],
@@ -213,7 +223,7 @@ from .fixtures import DandiAPI, SampleDandiset
             marks=mark.skipif_no_network,
         ),
         pytest.param(
-            "https://gui.dandiarchive.org/#/dandiset/001001/draft/files"
+            "https://dandiarchive.org/#/dandiset/001001/draft/files"
             "?location=sub-RAT123%2F",
             AssetFolderURL(
                 instance=known_instances["dandi"],
@@ -322,17 +332,6 @@ from .fixtures import DandiAPI, SampleDandiset
                 version_id=None,
             ),
         ),
-        pytest.param(
-            "https://gui.dandiarchive.org/#/dandiset/001001/draft/files"
-            "?location=sub-RAT123/*.nwb",
-            AssetFolderURL(
-                instance=known_instances["dandi"],
-                dandiset_id="001001",
-                version_id="draft",
-                path="sub-RAT123/*.nwb/",
-            ),
-            marks=mark.skipif_no_network,
-        ),
         (
             "dandi://dandi-api-local-docker-tests/000002/f*/bar.nwb",
             AssetItemURL(
@@ -370,17 +369,6 @@ def test_parse_api_url(url: str, parsed_url: ParsedDandiURL) -> None:
                 version_id="draft",
                 path="sub-RAT123/*.nwb",
             ),
-        ),
-        pytest.param(
-            "https://gui.dandiarchive.org/#/dandiset/001001/draft/files"
-            "?location=sub-RAT123/*.nwb",
-            AssetGlobURL(
-                instance=known_instances["dandi"],
-                dandiset_id="001001",
-                version_id="draft",
-                path="sub-RAT123/*.nwb",
-            ),
-            marks=mark.skipif_no_network,
         ),
         (
             "dandi://dandi-api-local-docker-tests/000002/f*/bar.nwb",

